QL: Update the use of some user-caused exceptions

bpintea opened the pull request:
This updates the use of certain exceptions in the QL, EQL and SQL code base. The failures that occur due to either user-provided input (such as incorrect query, incorrect query parameters), or data format not appropriate to the attempted operation (like converting strings to various other types) will now result in a 4xx type code, rather than 5xx.

This change also proposes turning `QlClientException` concrete, by implementing the `status()` method to return a `BAD_REQUEST` (400). The 400 is a "generic", catch-all client-side error message. The other 4xx errors are specializations. No QL code / exception makes use of specializations, but if they so wish, they can simply override the method.
This also makes it clearer to any subclass that a 4xx status must be return.

This change allows the use of some currently abstract subclasses, such as `EqlClientException`, which can be employed on user-related failures, that are currently inappropriately raised as a form of `QlIllegalArgumentException`, which is a 500-type error.

Also, an `InvalidArgumentException` extending `QlClientException` is introduced, being thrown where the execution cannot continue due to user's query (either direct input or invalid application of functions to index data).

bpintea committed:
Update EQL use of some exceptions

This updates the use of certain exceptions in the E/QL base. The failures
that occur due to either user-provided input (such as incorrect query,
incorrect query parameters), or data format not appropriate to the
attempted operation (like converting strings to various other types)
will now result in a 4xx type code, rather than 5xx.

This change also proposes turning `QlClientException` concrete, by
implementing the `status()` method to return a BAD_REQUEST (400).
The 400 is a "generic", catch-all client-side error message. The other
4xx errors are specializations. No QL code / exception makes use of
specializations, but if they so wish, they can simply override the
method.
This also makes it clearer to any subclass that a 4xx status
must be return; this isn't the case now,  `QlClientException` extending
`QlException` which is the root of both client and error exceptions.
.
This change allows the use of some currently abstract subclasses, such
as `EqlClientException`, which can be employed on user-related failures,
that are currently inappropriately raised as a form of
`QlIllegalArgumentException`, which is a 500-type error.

bpintea commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/util/StringUtils.java:
```diff
@@ -29,11 +26,4 @@ public static LikePattern toLikePattern(String s) {
         return new LikePattern(likeString, escape);
     }
-
-    public static LikePattern toLikePattern(Expression expression) {
```
not strictly required by change, but had to be inspected and it's not used anywhere.

alex-spies replied:
++, good find!

bpintea committed:
checkstyle police

remove import

alex-spies commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/EqlClientException.java:
```diff
@@ -8,16 +8,12 @@
 import org.elasticsearch.xpack.ql.QlClientException;
 
-public abstract class EqlClientException extends QlClientException {
+public class EqlClientException extends QlClientException {
```
thought: this will make the exception show in the REST response as `"type": "eql_client_exception"`. Ideally, we probably want to prefer slightly more specific subclasses of this (`VerificationException` is fine, for instance), but IMO this is a good solution for now.

bpintea replied:
The `VerificationException` is quite specific, though, requiring a `Failure`s collection, each requiring a `Source` which isn't available everywhere.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Concat.java:
```diff
@@ -103,12 +102,7 @@ private static int checkedTotalLength(BytesRef[] values) {
-                    return RestStatus.BAD_REQUEST; // return a 400 response
-                }
-            };
+            throw new EsqlClientException("concatenating more than [" + MAX_CONCAT_LENGTH + "] bytes is not supported");
```
praise: ++ for aligning this!

alex-spies commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/expression/function/scalar/string/CIDRMatchFunctionProcessorTests.java:
```diff
@@ -36,8 +36,8 @@ public void testCIDRMatchFunctionInvalidInput() {
         // Invalid source address
-        EqlIllegalArgumentException e = expectThrows(
-            EqlIllegalArgumentException.class,
+        Exception e = expectThrows(
```
nit: for the updated test cases in this file, I think throwing and expecting a verification exception would be more precise but this is fine as well.

alex-spies commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/expression/function/scalar/math/ToNumberFunctionProcessorTests.java:
```diff
@@ -19,12 +21,17 @@ private static Object process(Object value, Object base) {
+        return error(value, base, QlIllegalArgumentException.class);
+    }
+
+    private static String clientError(Object value, Object base) {
+        return error(value, base, EqlClientException.class);
     }
```
nit: Similarly here, we could also require a verification exception since all usages of `clientError` fall under this category IMO.

alex-spies commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java:
```diff
@@ -339,7 +339,7 @@ public Sequence visitSequence(SequenceContext ctx) {
 
         if (queries.stream().allMatch(KeyedFilter::isMissingEventFilter)) {
-            throw new IllegalStateException("A sequence requires at least one positive event query; found none");
+            throw new ParsingException("A sequence requires at least one positive event query; found none");
```
super nitpick: I think parsing went fine, so maybe verification exception would be more appropriate?

bpintea replied:
The `ParsingException` wouldn't have been 1st choice, but that's used above, so I reused it for consistency.

alex-spies commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/assembler/ExecutionManager.java:
```diff
@@ -177,7 +178,7 @@ public Executable assemble(
      */
     public Executable assemble(List<List<Attribute>> listOfKeys, List<PhysicalPlan> plans, Limit limit) {
         if (cfg.fetchSize() > SAMPLE_MAX_PAGE_SIZE) {
-            throw new EqlIllegalArgumentException("Fetch size cannot be greater than [{}]", SAMPLE_MAX_PAGE_SIZE);
+            throw new EqlClientException("Fetch size cannot be greater than [{}]", SAMPLE_MAX_PAGE_SIZE);
         }
```
nit: maybe also `VerificationException` here.

bpintea replied:
I've updated this usage, but with a newly introduced `QlClientIllegalArgumentException`. This is somewhat more fitting and this type of exception is required by the generic QL code changes (which this PR grew with).

alex-spies reviewed (approved):
Nice! Thanks @bpintea , this looks good. I expect this was quite arduous as well.

I only have optional remarks, mostly about usage of generic exceptions - but there's no right answer here, so please treat this as an opinion, mostly :)

bpintea committed:
Merge branch 'main' into eql/correct_exceptions_use

bpintea committed:
Update QL's QlIllegalArgumentException usage

Update QL's QlIllegalArgumentException usage

bpintea committed:
Merge branch 'main' into eql/correct_exceptions_use

bpintea committed:
Propagate QL changes to SQL and ESQL

Propagate QL changes to SQL and ESQL

alex-spies commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/QlClientIllegalArgumentException.java:
```diff
@@ -0,0 +1,23 @@
+ * Exception thrown when unable to continue processing client request,
+ * in cases such as invalid query or failure to apply requested processing to given data.
+ */
+public class QlClientIllegalArgumentException extends QlClientException {
```
bikeshed: we could also have a generic `InvalidQueryException` (and make `VerifierException` etc. inherit from it)

bpintea replied:
Thanks.
I've updated it to a "middle" `InvalidArgumentException` which is what all updates turned out to refer to, as QL-specific client-side version of `QlIllegalArgumentException`. I think this makes more sense as a user-facing `"type"`.

Also, I think a specialisation isn't actually needed; i.e. no `EqlInvalidArgumentException` or `SqlInvalidArgumentException`, like in the case of `QlIllegalArgumentException`(whose specialisations aren't very clear to me -- left a couple of TODOs).

alex-spies reviewed (approved):
Just re-reviewed, still LGTM!

bpintea committed:
Update exceptions use in SQL

This updates the 500-generating exceptions in SQL code.

Also, s/QlClientIllegalArgumentException/InvalidArgumentException as a
clearer type presented to the user.

bpintea committed:
Spotless applied

Spotless applied

bpintea committed:
Update new exception name in tests

Update tests that check on the exception name's being thrown.

bpintea committed:
Merge branch 'main' into eql/correct_exceptions_use

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/ExpressionBuilder.java:
```diff
@@ -117,7 +117,7 @@ public Literal visitIntegerValue(EsqlBaseParser.IntegerValueContext ctx) {
 
         try {
             number = StringUtils.parseIntegral(text);
-        } catch (QlClientIllegalArgumentException siae) {
```
super-duper nit: variable name of the exception is slightly less clear now (since it referred to a type); maybe let's just replace it by `e`.

bpintea replied:
Ah, yes, I'll update this with a next PR.

alex-spies commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/InvalidArgumentException.java:
```diff
@@ -0,0 +1,27 @@
+ * TODO: the intended use of ql.ParsingException, vs its [E|S|ES]QL equivalents, subclassed from the respective XxxClientException?
+ *       Same for PlanningException.
+ */
+public class InvalidArgumentException extends QlClientException {
```
++ on the nice javadoc, great!

nit: One unfortunate thing is that this might be confused with the built-in IAE. Maybe renaming to `InvalidClientArgumentException`, `InvalidUserArgumentException` or just `InvalidQueryException` would avoid that, but it's also fine as-is.

alex-spies commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTruncProcessor.java:
```diff
@@ -72,10 +69,10 @@ public static Object process(Object truncateTo, Object timestamp, ZoneId zoneId)
             && timestamp instanceof IntervalYearMonth == false
             && timestamp instanceof IntervalDayTime == false) {
-            throw new SqlIllegalArgumentException("A date/datetime/interval is required; received [{}]", timestamp);
+            throw new SqlIllegalArgumentException("A date/datetime/interval is required; received [{}]", timestamp); // verifier checked
```
Shouldn't this also be `InvalidArgumentException`?

bpintea replied:
The verifier checks this condition, so if it still gets here, it's a server error.

alex-spies commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/util/Check.java:
```diff
@@ -26,12 +27,6 @@ public static void isTrue(boolean expression, String message) {
         }
     }
 
-    public static void notNull(Object object, String message) {
```
Nice. Deleted code is good code.

alex-spies reviewed (approved):
Re-reviewed the newest changes.

I think this propels our exception hierarchy in the right direction. IMHO we can probably safely merge this without much further ado and put follow-up changes into separate PRs, if needed.

luigidellaquila reviewed (approved):
LGTM, thanks @bpintea !

bpintea commented:
Thanks, Alex, Luigi!

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/util/MathUtils.java:
```diff
@@ -14,6 +14,11 @@ public class MathUtils {
     public static int abs(int number) {
 
         if (number == Integer.MIN_VALUE) {
+            // TODO: can this function be removed?
```
Given the code that uses this method I would say it's not possible, but it's not so easy to evaluate this because there are some additions that happen. I would keep this code.

Nevertheless, I've went ahead and did some tests and there are some issues with code that sits before this method, more specifically `head` and `tail` pipes. I've created an issue [here](https://github.com/elastic/elasticsearch/issues/101724).

bpintea replied:
> Given the code that uses this method I would say it's not possible, but it's not so easy to evaluate this because there are some additions that happen.

When I looked over this, I thought that if we switched those additions to exact ones, this check would no longer be needed (and thus nor the method). In any case, for the purpose of this PR, keeping the exception as is in place makes sense.

Thanks for checking!

astefan reviewed (approved):
LGTM
